src/legend/Legend.component.js

```diff
@@ -77,14 +77,22 @@ class Legend extends Component {
         this.props.onItemsChange(newItems);
     };
 
-    updateItem = (newItems) => {
-        const modelToUpdate = legendItemStore.getState() && legendItemStore.getState().model;
-        const isNewLegendItem = newItems.every(model => model !== modelToUpdate);
-
-        return this.props.onItemsChange([].concat(
-            newItems,
-            isNewLegendItem ? modelToUpdate : [],
-        ));
+    updateItem = (existingItems) => {
+        let updatedItems = [];
+
+        const currentItem = legendItemStore.getState() && legendItemStore.getState().model;
+        const isNewItem = existingItems.every(item => item.id !== currentItem.id);
+
+        if (isNewItem) {
+            updatedItems = [...existingItems, currentItem];
+        } else {
+            updatedItems = existingItems.map((item) => {
+                const shouldBeUpdated = item.id === currentItem.id;
+                return shouldBeUpdated ? currentItem : item;
+            });
+        }
```

Comment:
Actually, you could rewrite this whole section a bit more clearly (and fewer loops through the array) as:

```
let updatedItems = existingItems;
const currentItem = legendItemStore.getState() && legendItemStore.getState().model

// This is necessary if the first half of the above expression ever evaluates to false,
// which it seems we expect it might?  Otherwise `currentItem.id` below will throw an error.
if (currentItem) {
  updatedItems = [...existingItems];
  const idx = updatedItems.findIndex(item => item.id === currentItem.id)
  if (idx === -1) {
    updatedItems.push(currentItem)
  } else {
    updatedItems[idx] = currentItem
  }
}

return this.props.onItemsChange(updatedItems);
```

Not a required change, just thoughts while thinking

Reply:
Nice, though I'm not sure about `findIndex` as we might eventually decide to use only current syntax. Your addition of a failsafe for a missing model is a good thing to do anyway, I'll add that, thanks!

Reply:
Though on the other hand currently we're using stage-0 with babel, so maybe that's a bit too careful.

Reply:
findIndex is in the ES2015 standard so I think it's fine to use.  We need to polyfill things for IE anyway, just worth making sure we're doing that :-)

Reply:
You're right, I've added your suggested changes. Thanks!